### DISCUSSION TITLE
In Memory Database

### DISCUSSION THREAD

**Discussion by satoshi-szk** [powersync-ja/sqlite_async.dart#13] [Ideas]
Hi, thank you for your great library.
I am really happy because I was looking for a library that would manage Isolate and allow concurrent transactions.

Anyway, I often use in-memory databases for unit testing. I think it will be useful for many people.
The sqlite3 library has a method `sqlite3.openInMemory()` available; the SqliteDatabase class in sqlite_async does not appear to have such an option.
Please let me know if I am just missing something that is already implemented.

**Comment by rkistner** [comment]
The current implementation is optimized for concurrent transactions, which is not supported with in-memory SQLite databases. We could expose the same API without concurrency with an in-memory database, but that would require some work. Can investigate that for a future release.

**Reply by satoshi-szk** [reply]
Thanks for the info. 
If the behavior changes between in-memory mode and normal mode, it may be preferable to run in normal mode for unit test use cases like mine.

**Reply by rkistner** [reply]
Personally, I've also found it useful to be able to inspect the database when tests fail. For the library itself, I use a separate database file per test.

You can see the implementation here:

https://github.com/journeyapps/sqlite_async.dart/blob/d937c5d6d2b1ee4f22a2e14311d614f869087e5b/test/basic_test.dart#L13-L32

https://github.com/journeyapps/sqlite_async.dart/blob/d937c5d6d2b1ee4f22a2e14311d614f869087e5b/test/util.dart#L89-L97

**Reply by satoshi-szk** [reply]
Thank you. I gave it a shot, and this has definitely been very helpful for me.
I believe it might be worth considering including these utilities in the library, so that users can simply import them.

**Reply by ngoc-quoc-huynh** [reply]
@rkistner Would it be possible to expose these methods, so we don't need to copy paste them?

**Reply by rkistner** [reply]
> @rkistner Would it be possible to expose these methods, so we don't need to copy paste them?

The code uses some internals of the test library and uses conventions specific to the tests here, so I'm not sure if it's feasible to expose as re-usable code.
